### ISSUE TITLE
Kafka indexing gets stuck if there is data from outside of load period

### ISSUE THREAD

**Issue by gvsmirnov** [apache/incubator-druid#4137]
**Steps to reproduce:**

1. Set up a druid cluster with `druid-kafka-indexing-service` enabled, using the [sample spec provided in the doc](http://druid.io/docs/latest/development/extensions-core/kafka-ingestion.html#submitting-a-supervisor-spec)

2. Send data to kafka:
```
$ cat current.data.json
{ "timestamp": "2017-03-30T00:00:00Z", "value": 42.0 }

$ cat current.data.json | kafka/bin/kafka-console-producer.sh --broker-list localhost:9092 --topic metrics
```

3. Set up a retention rule that asks to load data for the last day:
<img width="589" alt="screen shot 2017-03-31 at 00 40 49" src="https://cloud.githubusercontent.com/assets/1493464/24527479/fd0731da-15aa-11e7-94db-94cabf5319e2.png">

4. Send outdated data to kafka:
```
$ cat old.data.json 
{ "timestamp": "2016-04-30T00:00:00Z", "value": 42.0 }

$ cat old.data.json | kafka/bin/kafka-console-producer.sh --broker-list localhost:9092 --topic metrics
```

5. Wait for the task to complete (takes an hour by default, can be changed in `ioConfig.taskDuration` of the `supervisor-spec.json` file)

**Expected behavior:**
Current task completes right away, a new task is started

**Actual behavior:**
The task hangs while trying to hand off the segment:

```
2017-03-30T21:38:29,659 INFO [task-runner-0-priority-0] io.druid.segment.realtime.appenderator.FiniteAppenderatorDriver - Awaiting handoff of segments: [metrics-kafka_2016-04-30T00:00:00.000Z_2016-04-30T01:00:00.000Z_2017-03-30T21:36:45.791Z]
2017-03-30T21:39:28,499 INFO [coordinator_handoff_scheduled_0] io.druid.segment.realtime.plumber.CoordinatorBasedSegmentHandoffNotifier - Still waiting for Handoff for Segments : [[SegmentDescriptor{interval=2016-04-30T00:00:00.000Z/2016-04-30T01:00:00.000Z, version='2017-03-30T21:36:45.791Z', partitionNumber=0}]]
2017-03-30T21:40:28,479 INFO [coordinator_handoff_scheduled_0] io.druid.segment.realtime.plumber.CoordinatorBasedSegmentHandoffNotifier - Still waiting for Handoff for Segments : [[SegmentDescriptor{interval=2016-04-30T00:00:00.000Z/2016-04-30T01:00:00.000Z, version='2017-03-30T21:36:45.791Z', partitionNumber=0}]]
2017-03-30T21:41:28,478 INFO [coordinator_handoff_scheduled_0] io.druid.segment.realtime.plumber.CoordinatorBasedSegmentHandoffNotifier - Still waiting for Handoff for Segments : [[SegmentDescriptor{interval=2016-04-30T00:00:00.000Z/2016-04-30T01:00:00.000Z, version='2017-03-30T21:36:45.791Z', partitionNumber=0}]]
2017-03-30T21:42:28,478 INFO [coordinator_handoff_scheduled_0] io.druid.segment.realtime.plumber.CoordinatorBasedSegmentHandoffNotifier - Still waiting for Handoff for Segments : [[SegmentDescriptor{interval=2016-04-30T00:00:00.000Z/2016-04-30T01:00:00.000Z, version='2017-03-30T21:36:45.791Z', partitionNumber=0}]]
2017-03-30T21:43:28,479 INFO [coordinator_handoff_scheduled_0] io.druid.segment.realtime.plumber.CoordinatorBasedSegmentHandoffNotifier - Still waiting for Handoff for Segments : [[SegmentDescriptor{interval=2016-04-30T00:00:00.000Z/2016-04-30T01:00:00.000Z, version='2017-03-30T21:36:45.791Z', partitionNumber=0}]]
```

**Comment by pjain1** [comment]
do you have any default rule setup ? I think the problem is that for the outdated segment default rule will be applicable since the segment it is not within the last day and if the default rule not a load rule then the segment may not be loaded and task will not finish.

**Comment by gvsmirnov** [comment]
@pjain1 the only rule change from default is in the ticket. The default rule's load was changed from forever to `P1D`.

**Comment by pjain1** [comment]
hmm..tasks can be configured to drop such outdated message see `lateMessageRejectionPeriod ` in `KafkaSupervisorIOConfig` section [here](http://druid.io/docs/latest/development/extensions-core/kafka-ingestion.html). Assumption here is that If the late message will be indexed by batch pipeline eventually.

**Comment by iNikem** [comment]
But this is counterintuitive, that Historical node's load rule breaks segment hand-off from indexing node. I mean, one would assume that the responsibility of indexing node is to prepare the segment ready and signal coordinator about it. If some or any historicals refuse to load that segment due to their rules, well, then this segment is not currently needed and it can live quietly in deep storage. Why block the hand-off?

**Comment by pjain1** [comment]
The current design makes sure that the indexed is always available for querying therefore task do not shutdown unless the segment is loaded by a historical. 

However, you are correct, in an ideal world task can drop segments if its not going to get loaded. Currently the indexing tasks does not have any knowledge about load/drop rules so they can't make this decision. Hopefully, this can be improved in a future version, if you are interested you are welcome to do a PR :) I can help you with it.

Another argument that can be made is that always set `lateMessageRejectionPeriod` as per the load/drop rules so that late data is dropped as anyways it is not going to get loaded.

**Comment by gianm** [comment]
I think this problem would be addressed by #5868 and #5869. Is anyone affected by this interested in working on one or both of those?

**Comment by xy-xin** [comment]
We encounter this problem recently. Will kafka-indexing-service skip these data after some retries?

**Comment by gianm** [comment]
@xy-xin I think yes, although it is not super graceful (they will have to time out first). You could set `lateMessageRejectionPeriod` as a workaround until #5868 and #5869 are implemented (I think they are the ultimate right approach to this).

**Comment by xy-xin** [comment]
Thanks for your reply, @gianm . Have confirmed the result, and nothing bad impact. But still looking forward to #5868  and #5869

**Comment by jihoonson** [comment]
I think this is fixed in #5868 and #5869. Closing now.
